provider/mesh/state.go

```diff
@@ -70,6 +52,47 @@ func (st *notificationState) gc(retention time.Duration) {
 	}
 }
 
+func (st *notificationState) snapshot(w io.Writer) error {
+	st.mtx.RLock()
+	defer st.mtx.RUnlock()
+
+	enc := gob.NewEncoder(w)
+	for k, n := range st.set {
+		if err := enc.Encode(k); err != nil {
+			return err
+		}
+		if err := enc.Encode(n); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (st *notificationState) loadSnapshot(r io.Reader) error {
+	st.mtx.Lock()
+	defer st.mtx.Unlock()
+
+	dec := gob.NewDecoder(r)
+	for {
+		var n notificationEntry
+		var k string
+		if err := dec.Decode(&k); err != nil {
+			// Only EOF at the start of new pair is correct.
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+		if err := dec.Decode(&n); err != nil {
+			return err
+		}
+		// TODO(fabxc): remove serialization of alert/receiver into
+		// string key.
+		st.set[k] = n
```

Comment:
Serialization of state into messages for gossip communication and snapshotting is completely independent. Right now it's mostly the same (modulo your comment below) but I intend the encoding of either.

Think of this PR more to be about everything but the exact format.